server/workflowtemplate/workflow_template_server.go

```diff
@@ -67,17 +69,87 @@ func (wts *WorkflowTemplateServer) getTemplateAndValidate(ctx context.Context, n
 	return wfTmpl, nil
 }
 
+func cursorPaginationByResourceVersion(items []v1alpha1.WorkflowTemplate, resourceVersion string, limit int64, wfList *v1alpha1.WorkflowTemplateList) {
+	// Sort the workflow list in descending order by resourceVersion.
+	sort.Slice(items, func(i, j int) bool {
+		itemIRV, _ := strconv.Atoi(items[i].ResourceVersion)
+		itemJRV, _ := strconv.Atoi(items[j].ResourceVersion)
+		return itemIRV > itemJRV
+	})
+
+	// resourceVersion: unique value to identify the version of the object by Kubernetes. It is used for pagination in workflows.
+	// receivedRV: resourceVersion value used for previous pagination
+	// Due to the descending sorting above, the items are filtered to have a resourceVersion smaller than receivedRV.
+	// The data with values smaller than the receivedRV on the current page will be used for the next page.
+	if resourceVersion != "" {
+		var newItems []v1alpha1.WorkflowTemplate
+		for _, item := range items {
+			targetRV, _ := strconv.Atoi(item.ResourceVersion)
+			receivedRV, _ := strconv.Atoi(resourceVersion)
+			if targetRV < receivedRV {
+				newItems = append(newItems, item)
+			}
+			items = newItems
+		}
+	}
+
+	// Indexing list by limit count
+	if limit != 0 {
+		endIndex := int(limit)
+		if endIndex > len(items) || limit == 0 {
+			endIndex = len(items)
+		}
+		wfList.Items = items[0:endIndex]
+	} else {
+		wfList.Items = items
+	}
+
+	// Calculate new offset for next page
+	// For the next pagination, the resourceVersion of the last item is set in the Continue field.
+	if limit != 0 && len(wfList.Items) == int(limit) {
+		lastIndex := len(wfList.Items) - 1
+		wfList.ListMeta.Continue = wfList.Items[lastIndex].ResourceVersion
+	}
+}
+
 func (wts *WorkflowTemplateServer) ListWorkflowTemplates(ctx context.Context, req *workflowtemplatepkg.WorkflowTemplateListRequest) (*v1alpha1.WorkflowTemplateList, error) {
 	wfClient := auth.GetWfClient(ctx)
-	options := &v1.ListOptions{}
+	k8sOptions := &v1.ListOptions{}
+
 	if req.ListOptions != nil {
-		options = req.ListOptions
+		k8sOptions = req.ListOptions
+	}
+
+	// Save the original Continue and Limit for custom filtering.
+	resourceVersion := k8sOptions.Continue
+	limit := k8sOptions.Limit
+
+	if req.NamePattern != "" {
+		// Search whole with Limit 0.
+		// Reset the Continue "" to prevent Kubernetes native pagination.
+		// Kubernetes api will search for all results without limit and pagination.
+		k8sOptions.Continue = ""
+		k8sOptions.Limit = 0
```

Comment:
correct me if I'm wrong, but won't this have the same problem as #11761 / i.e. #12025 when there are enough `WorkflowTemplates`? It looks like the exact problem was mentioned in https://github.com/argoproj/argo-workflows/issues/11004#issuecomment-1690298225

Reply:
The workaround here may be to only retrieve the `limit` at a time from the k8s API. If there aren't enough after the filter, then retrieve more.

That might make the request take longer (potentially significantly, but maybe less than retrieving all at once, which also takes time, and, crucially, memory), but would make sure we never retrieve too many (i.e. it throttles it, effectively). There are some other side-effects to that too (the k8s API rate limit, for instance).

The other way may be to do this fully client-side or partially client-side.
Fully would be the UI (and CLI? a flag wasn't implemented here though) fetches and does a filter, and then if it doesn't have enough, requests more.
Partially would be the API does a filter but can underfetch, in which case the UI has to request more.

Reply:
I'm also thinking we may want to explicitly point out in the UI which search fields are not k8s native and can therefore be non-performant when used; basically a "hey we implemented this, but you might want to be careful using this".

and at that point, potentially let admins disable them with flags too, if they don't want users to overload the Argo Server or k8s API accidentally.

Reply:
For future reference, the current pattern to mitigate this would be to follow the SQLite pattern here a la #12736, see also https://github.com/argoproj/argo-workflows/pull/13633#discussion_r1776448455 et al